message-passing/safe-message-handlers/src/cluster-manager.ts

```diff
@@ -37,27 +37,39 @@ export class ClusterManager {
   }
 
   async startCluster(): Promise<void> {
-    await startCluster();
-    this.state.clusterStarted = true;
-    for (let i = 0; i < 25; i++) {
-      this.state.nodes.set(i.toString(), null);
+    if (this.state.clusterState !== ClusterState.NOT_STARTED) {
+      // This is used as a Signal handler so we log a warning but cannot return an error.
+      wf.log.warn(`Cannot start cluster in state ${this.state.clusterState}`);
+      return;
     }
+    await this.nodesMutex.runExclusive(async () => {
+      await startCluster();
+      this.state.clusterState = ClusterState.STARTED;
+      for (let i = 0; i < 25; i++) {
+        this.state.nodes.set(i.toString(), null);
+      }
+    });
     wf.log.info('Cluster started');
   }
 
-  async shutDownCluster(): Promise<void> {
-    await wf.condition(() => this.state.clusterStarted);
-    this.state.clusterShutdown = true;
+  async shutdownCluster(): Promise<true> {
+    if (this.state.clusterState !== ClusterState.STARTED) {
+      // This is used as an Update handler we return an error to the caller.
+      throw new wf.ApplicationFailure(`Cannot shutdown cluster in state ${this.state.clusterState}`);
+    }
+    await shutdownCluster();
+    this.state.clusterState = ClusterState.SHUTTING_DOWN;
     wf.log.info('Cluster shutdown');
+    return true;
   }
 
   async assignNodesToJob(input: AssignNodesToJobUpdateInput): Promise<ClusterManagerStateSummary> {
-    await wf.condition(() => this.state.clusterStarted);
-    if (this.state.clusterShutdown) {
+    await wf.condition(() => this.state.clusterState === ClusterState.STARTED);
```

Comment:
@dandavison I just noticed is there a deadlock here if the `clusterState` goes from `STARTED` to `SHUTTING_DOWN` without this handler running?